packages/core/vue/src/useWallet.ts

```diff
@@ -0,0 +1,225 @@
+    // Methods.
+    select(walletName: WalletName): void;
+    connect(): Promise<void>;
+    disconnect(): Promise<void>;
+    sendTransaction(transaction: Transaction, connection: Connection, options?: SendTransactionOptions): Promise<TransactionSignature>;
+    signTransaction(transaction: Transaction): Promise<Transaction>;
+    signAllTransactions(transaction: Transaction[]): Promise<Transaction[]>;
+    signMessage(message: Uint8Array): Promise<Uint8Array>;
+}
+
+export interface WalletStoreProps {
+    wallets: Wallet[];
+    autoConnect?: boolean;
+    onError?: (error: WalletError) => void;
+    localStorageKey?: string;
+}
+
+let walletStore: WalletStore = {} as WalletStore;
+
+export const useWallet = (): WalletStore => walletStore;
+
+export const initWallet = ({
+    wallets,
+    autoConnect = false,
+    onError = (error: WalletError) => console.error(error),
+    localStorageKey = 'walletName',
+}: WalletStoreProps): void => {
+    const walletName = useLocalStorage<string>(localStorageKey);
+    const wallet = ref<Wallet | null>(null);
+    const adapter = ref<Adapter | null>(null);
+    const publicKey = ref<PublicKey | null>(null);
+    const ready = ref<boolean>(false);
+    const connected = ref<boolean>(false);
+    const connecting = ref<boolean>(false);
+    const disconnecting = ref<boolean>(false);
+
+    const walletsByName = computed<WalletDictionary>(() => {
+        return wallets.reduce((walletsByName, wallet) => {
+            walletsByName[wallet.name] = wallet;
+            return walletsByName;
+        }, {} as WalletDictionary);
+    });
+
+    // Update the wallet and adapter based on the wallet provider.
+    watch(walletName, (): void => {
+        wallet.value = walletsByName.value?.[walletName.value as string] ?? null;
+        adapter.value = wallet.value?.adapter() ?? null;
+        if (adapter.value) {
+            ready.value = adapter.value.ready;
+            publicKey.value = adapter.value.publicKey;
+            connected.value = adapter.value.connected;
+        } else {
+            ready.value = false;
+            publicKey.value = null;
+            connected.value = false;
+        }
+    }, { immediate:true });
+
+    // Select a wallet by name.
+    const select = async (newWalletName: string): Promise<void> => {
+        if (walletName.value === newWalletName) return
+        if (adapter.value) await adapter.value.disconnect()
+        walletName.value = newWalletName
+    };
+
+    // Handle the adapter events.
+    const onReady = () => ready.value = true;
+    const onConnect = () => {
+        if (! adapter.value) return;
+        ready.value = adapter.value.ready;
+        publicKey.value = adapter.value.publicKey;
+        connected.value = adapter.value.connected;
+    };
+    watchEffect(onInvalidate => {
+        if (! adapter.value) return;
+        adapter.value.on('ready', onReady);
+        adapter.value.on('connect', onConnect);
+        adapter.value.on('error', onError);
```

Comment:
What about `disconnect`?

Reply:
I had issues with the `onDisconnect` callback when defined as `() => walletName.value = null`.

When using the `autoconnect` flag, it would cause the whole thing to disconnect on refresh.

The flow was:
- Connect your wallet using the `autoconnect` flag.
- Refreshing the page would call `onDisconnect` on the adapter.
- On reload, the walletName would be empty and so we wouldn't be able to autoconnect.

Additionally, we explicitly set `walletName.value = null` in the `finally` scope of the `disconnect` method so I thought this was perhaps redundant.

I'm happy to give it another go though, did you come across that issue with React?

Reply:
> Refreshing the page would call onDisconnect on the adapter.

This one is interesting, and seems likely to be the cause. I wonder why that is, as I don't think React does this?

Reply:
A _possible_ workaround could be to add an `onbeforeunload` event listener to unregister the `disconnect` event handler so this doesn't happen. But we do need a `disconnect` event listener, since adapters can disconnect for various reasons outside the control of the app.

Reply:
It worked! I understand what was happening now.

Just like in React, the `onInvalidate` callback (that disables all adapter listeners) defined in the `watchEffect` method will be called when the component is unmounted. From the [documentation](https://v3.vuejs.org/guide/reactivity-computed-watchers.html#side-effect-invalidation):

> This invalidation callback is called when:
>
> - the effect is about to re-run
> - the watcher is stopped (i.e. when the component is unmounted if watchEffect is used inside setup() or lifecycle hooks)

However, in the `useWallet.ts` file, we're not inside a particular component as we're defining a global store. In React, it is more common to define these in `ContextProviders` which are components themselves giving data to their descendants. In Vue, we have something similar with `provide` and `inject` but it's not quite the same and is rarely used in the community.

So we can do two things to go around this:
- Since a component will ultimately be calling the `initWallet` method, we could ask for a component as a parameter and then give it to the `watchEffect` method so it hooks itself to whenever that component unmounts. The only issue with this approach is we're dirtying a little bit the API by having to ask the component to add itself to the parameters.
- Do what you suggested and add a `beforeunload` window listener that invalidates the adapter listeners.

Both of these make sure adapter listeners are invalidated before the `onDisconnect` callback is triggered by the wallet itself upon refresh.

I went for the second approach to keep the API cleaner and less coupled to components. Do let me know if you'd rather I went for the first approach (or any other approach).

Thanks for your help in this it was really bugging me and I feel much better now that I understand what's happening haha.

Reply:
My concern about the `beforeunload` event is that I don't know if it could result in a race condition. For example, if the wallet's `disconnect` is triggered by `beforeunload` or something, and that event handler is called before the one here. I'm not saying we shouldn't do this, but it's something to be aware of, and could be nasty to debug if reported by users.

Reply:
We could maybe add a `cleanup`/`teardown` method to the store, and then a top level component unmounting could call it? It's guaranteed that devs will forget to do this though...